src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -729,153 +728,277 @@ func (s *service) FetchTagged(tctx thrift.Context, req *rpc.FetchTaggedRequest)
+	}
+	for _, readers := range i.idResult.blockReaders {
+		segments, err := readEncodedResultSegment(readers)
 		if err != nil {
 			return err
 		}
-		tags := idxconvert.ToSeriesTags(metadata, idxconvert.Opts{NoClone: true})
-
-		enc := s.pools.tagEncoder.Get()
-		ctx.RegisterFinalizer(enc)
-		encodedTags, err := s.encodeTags(enc, tags)
-		if err != nil { // This is an invariant, should never happen
-			return tterrors.NewInternalError(err)
-		}
-
-		elem := &rpc.FetchTaggedIDResult_{
-			NameSpace:   nsIDBytes,
-			ID:          id.Bytes(),
-			EncodedTags: encodedTags.Bytes(),
-		}
-		response.Elements = append(response.Elements, elem)
-		if !fetchData {
+		if segments == nil {
 			continue
 		}
-
-		encoded, err := db.ReadEncoded(ctx, nsID, id,
-			opts.StartInclusive, opts.EndExclusive)
-		if err != nil {
-			return convert.ToRPCError(err)
-		} else {
-			encodedDataResults[idx] = encoded
-		}
+		i.segments = append(i.segments, segments)
 	}
 	return nil
 }
 
-func (s *service) fetchReadResults(
-	ctx context.Context,
-	response *rpc.FetchTaggedResult_,
-	nsID ident.ID,
-	encodedDataResults [][][]xio.BlockReader,
-) error {
-	ctx, sp, sampled := ctx.StartSampledTraceSpan(tracepoint.FetchReadResults)
-	if sampled {
-		sp.LogFields(
-			opentracinglog.String("id", nsID.String()),
-			opentracinglog.Int("elementCount", len(response.Elements)),
-		)
+type fetchTaggedResultsIter struct {
+	idIter         idIter
+	idResults      []*idResult
+	startInclusive time.Time
+	endExclusive   time.Time
+	db             storage.Database
+	idx            int
+	exhaustive     bool
+	fetchData      bool
+}
+
+type fetchTaggedResultsIterOpts struct {
+	queryResult *index.QueryResult
+	queryOpts   *index.QueryOptions
+	fetchData   bool
+	db          storage.Database
+	docReader   *docs.EncodedDocumentReader
+	nsID        ident.ID
+	tagEncoder  serialize.TagEncoder
+	iOpts       instrument.Options
+}
+
+type idResult struct {
+	queryResult  *index.ResultsMapEntry
+	blockReaders [][]xio.BlockReader
+}
+
+func newFetchTaggedResultsIter(opts *fetchTaggedResultsIterOpts) FetchTaggedResultsIter {
+	iter := &fetchTaggedResultsIter{
+		idResults:      make([]*idResult, 0, opts.queryResult.Results.Map().Len()),
+		exhaustive:     opts.queryResult.Exhaustive,
+		db:             opts.db,
+		fetchData:      opts.fetchData,
+		startInclusive: opts.queryOpts.StartInclusive,
+		endExclusive:   opts.queryOpts.EndExclusive,
+		idIter: idIter{
+			docReader:  opts.docReader,
+			nsID:       opts.nsID,
+			tagEncoder: opts.tagEncoder,
+			iOpts:      opts.iOpts,
+		},
 	}
-	defer sp.Finish()
 
-	for idx := range response.Elements {
-		segments, rpcErr := s.readEncodedResult(ctx, encodedDataResults[idx])
-		if rpcErr != nil {
-			return rpcErr
+	for _, entry := range opts.queryResult.Results.Map().Iter() { // nolint: gocritic
+		entry := entry
+		iter.idResults = append(iter.idResults, &idResult{
+			queryResult: &entry,
+		})
```

Comment:
Can we make both `idResult` and `index.ResultsMapEntry` not be using pointers here?

Both will cause these structs to be individually heap allocated (due to escape analysis unsure of the lifetimes of these two structs).

Using by value and non-pointer types will ensure that (1) idResult in the idResults slice can be just part of the allocation of the slice and (2) index.ResultsMapEntry won't need to be heap allocated and a memcpy can move the struct from the results map to this value.